core/src/worker/workflow/workflow_stream.rs

```diff
@@ -339,48 +341,36 @@ impl WFStream {
+        });
         // Delete the activation, but only if the report came from lang, or we know the outstanding
         // activation is expected to be completed internally.
-        if let Some(activation) = self.runs.get_mut(run_id).and_then(|rh| {
-            rh.delete_activation(|act| {
-                !report.is_autocomplete || matches!(act, OutstandingActivation::Autocomplete)
-            })
+        if let Some((should_evict, mut maybe_buffered)) = self.runs.get_mut(run_id).map(|rh| {
+            rh.finish_activation(
+                |act| !report.is_autocomplete || matches!(act, OutstandingActivation::Autocomplete),
+                wft_from_complete,
+            )
         }) {
-            // Evict the run if the activation contained an eviction
-            let mut applied_buffered_poll_for_this_run = false;
-            if activation.has_eviction() {
+            if should_evict {
                 debug!(run_id=%run_id, "Evicting run");
+                self.runs.remove(run_id);
+            }
 
-                if let Some(mut rh) = self.runs.remove(run_id) {
-                    if let Some(buff) = rh.take_buffered_wft() {
-                        // Don't try to apply a buffered poll for this run if we just got a new WFT
-                        // from completing, because by definition that buffered poll is now an
-                        // out-of-date WFT.
-                        if wft_from_complete.is_none() {
-                            res = self.instantiate_or_update(buff);
-                            applied_buffered_poll_for_this_run = true;
-                        }
-                    }
-                }
-
+            let maybe_buffered_wft = maybe_buffered
+                .get_next_wft()
                 // Attempt to apply a buffered poll for some *other* run, if we didn't have a wft
                 // from complete or a buffered poll for *this* run.
-                if wft_from_complete.is_none() && !applied_buffered_poll_for_this_run {
-                    if let Some(buff) = self.buffered_polls_need_cache_slot.pop_front() {
-                        res = self.instantiate_or_update(buff);
-                    }
-                }
-            };
-        }
-
-        if let Some((wft, pag)) = wft_from_complete {
-            debug!(run_id=%wft.execution.run_id, "New WFT from completion");
-            if let Some(t) = maybe_t {
-                res = self.instantiate_or_update(PermittedWFT {
-                    work: wft,
-                    permit: t.permit,
-                    paginator: pag,
-                });
+                .or_else(|| self.buffered_polls_need_cache_slot.pop_front());
+            if let Some(wft) = maybe_buffered_wft {
+                res = self.instantiate_or_update(wft);
+                // We accept that there might be tasks remaining in the buffer if we evicted and
+                // re-instantiated here. It's likely those tasks are now invalidated anyway.
```

Comment:
I could theoretically carry over any remaining buffered tasks to the next instantiation, but it adds quite a lot of complexity for little benefit IMO.